File: src/backend/src/controllers/console.controller.js

const fs = require('fs');
const fsPromises = require('fs').promises;
const path = require('path');

class ConsoleController {
    constructor(indexControllerInstance) {
        this.indexController = indexControllerInstance;
        console.log("ConsoleController instantiated and linked with IndexController.");
    }

    async getLiveConsole(req, res, next) {
        const { serverId } = req.params;
        const charOffset = parseInt(req.query.offset || '0', 10);

        const servers = this.indexController._readServers();
        const server = servers.find(s => s.id === serverId);

        if (!server) {
            return res.status(404).json({ message: "Server not found." });
        }

        const currentStatus = server.status; 

        if (currentStatus === 'Offline' || currentStatus === 'Stopping') { 
             return res.status(200).json({ logs: `\n--- Server is ${currentStatus}. Console inactive. ---\n`, newOffset: charOffset, status: currentStatus });
        }

        if (!server.consoleLogFile) {
            return res.status(404).json({ message: "Console log file path not configured for this server." });
        }

        try {
            if (!fs.existsSync(server.consoleLogFile)) {
                 return res.status(200).json({ logs: `--- Console log file not yet created or server not fully started. ---\n`, newOffset: 0, status: currentStatus });
            }

            const stats = await fsPromises.stat(server.consoleLogFile);
            const totalSize = stats.size;

            if (charOffset >= totalSize && currentStatus !== 'Starting' && currentStatus !== 'Online' && currentStatus !== 'Restarting') { 
                 return res.status(200).json({ logs: `\n--- Server is ${currentStatus}. Console may be stale. ---\n`, newOffset: totalSize, status: currentStatus });
            } else if (charOffset >= totalSize) {
                return res.status(200).json({ logs: '', newOffset: totalSize, status: currentStatus });
            }

            const bufferSize = totalSize - charOffset;
            const buffer = Buffer.alloc(bufferSize);
            const fileDescriptor = await fsPromises.open(server.consoleLogFile, 'r');

            await fileDescriptor.read(buffer, 0, bufferSize, charOffset);
            await fileDescriptor.close();

            const newContent = buffer.toString('utf-8');

            res.status(200).json({ logs: newContent, newOffset: totalSize, status: currentStatus });

        } catch (error) {
            console.error(`Error in ConsoleController.getLiveConsole for server ${serverId}:`, error);
            next(error);
        }
    }

    async getFullLog(req, res, next) {
        const { serverId } = req.params;
        try {
            const servers = this.indexController._readServers();
            const server = servers.find(s => s.id === serverId);

            if (!server) {
                return res.status(404).json({ message: "Server not found." });
            }

            if (!server.consoleLogFile || !fs.existsSync(server.consoleLogFile)) {
                return res.status(404).json({ message: "Console log file not found for this server. It may not have been run yet." });
            }

            const logContent = await fsPromises.readFile(server.consoleLogFile, 'utf-8');
            res.status(200).type('text/plain').send(logContent);

        } catch (error) {
            console.error(`Error in getFullLog for server ${serverId}:`, error);
            next(error);
        }
    }

    async sendCommandToServer(req, res, next) {
        const { serverId } = req.params;
        const { command } = req.body;

        if (!command || typeof command !== 'string' || !command.trim()) {
            return res.status(400).json({ message: "Command is required and must be a non-empty string." });
        }

        const servers = this.indexController._readServers();
        const server = servers.find(s => s.id === serverId);

        if (!server) {
            return res.status(404).json({ message: "Server not found." });
        }

        const allowedStatusesForCommand = ['Online', 'Starting', 'Restarting']; 
        if (!allowedStatusesForCommand.includes(server.status)) {
             return res.status(400).json({ message: `Cannot send command. Server is currently ${server.status}.` });
        }

        const serverProcess = this.indexController.activeServerProcesses[serverId];
        const trimmedCommand = command.trim();

        if (serverProcess && serverProcess.recovered === true && !serverProcess.stdin) {
            return res.status(400).json({ 
                message: "Cannot send command: This server was found running when the panel (re)started and its command input cannot be re-attached. To regain full command control, please 'Restart' the server using the panel's action buttons." 
            });
        }

        if (serverProcess && serverProcess.stdin && !serverProcess.stdin.destroyed) {
            try {
                serverProcess.stdin.write(trimmedCommand + '\n');
                console.log(`ConsoleController: Sent command "${trimmedCommand}" to server ${server.name} (ID: ${serverId}, PID: ${serverProcess.pid || 'N/A'})`);
                return res.status(200).json({ message: `Command "${trimmedCommand}" sent.` });
            } catch (error) {
                console.error(`Error writing command to server ${serverId} stdin (PID: ${serverProcess.pid || 'N/A'}, Status: ${server.status}, Command: "${trimmedCommand}"): ${error.message}`);
                next(error);
            }
        } else {
            let reason = "Server process not found in panel's active memory or its command input (stdin) is not available.";
            if (!serverProcess) {
                reason = `No active server process found in memory for server ${serverId} (Status: ${server.status}, PID from record: ${server.pid || 'N/A'}).`;
            } else if (!serverProcess.stdin) {
                 reason = `Active server process found for server ${serverId} (PID: ${serverProcess.pid || 'N/A'}), but its stdin is not available. This can happen if the panel restarted while the server was running.`;
            } else if (serverProcess.stdin.destroyed) {
                reason = `Active server process found for server ${serverId} (PID: ${serverProcess.pid || 'N/A'}), but its stdin is destroyed.`;
            }
             console.log(`ConsoleController: ${reason} Command: "${trimmedCommand}"`);
            return res.status(404).json({ message: reason });
        }
    }
}

module.exports = ConsoleController;

File: src/backend/src/controllers/console.controller.rs
use std::path::{Path, PathBuf};
use std::process::{Child, ChildStdin};
use std::sync::{Arc, Mutex};
use actix_web::{web, HttpResponse, Responder};
use tokio::fs::{self, File};
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use thiserror::Error;
use serde::{Serialize, Deserialize};

#[derive(Debug, Error)]
pub enum ConsoleError {
    #[error("Server not found")]
    ServerNotFound,
    #[error("Console log file not configured")]
    LogFileNotConfigured,
    #[error("Console log file not found")]
    LogFileNotFound,
    #[error("Invalid command")]
    InvalidCommand,
    #[error("Server not in correct state: {0}")]
    InvalidServerState(String),
    #[error("Process input not available")]
    ProcessInputUnavailable,
    #[error("IO error: {0}")]
    IoError(#[from] std::io::Error),
    #[error("Internal server error")]
    InternalError,
}

impl actix_web::error::ResponseError for ConsoleError {
    fn status_code(&self) -> actix_web::http::StatusCode {
        match self {
            ConsoleError::ServerNotFound | ConsoleError::LogFileNotFound => {
                actix_web::http::StatusCode::NOT_FOUND
            }
            ConsoleError::InvalidCommand | ConsoleError::InvalidServerState(_) => {
                actix_web::http::StatusCode::BAD_REQUEST
            }
            ConsoleError::ProcessInputUnavailable => actix_web::http::StatusCode::CONFLICT,
            _ => actix_web::http::StatusCode::INTERNAL_SERVER_ERROR,
        }
    }

    fn error_response(&self) -> actix_web::HttpResponse {
        HttpResponse::build(self.status_code())
            .json(serde_json::json!({ "message": self.to_string() }))
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Server {
    pub id: String,
    pub name: String,
    pub status: String,
    pub console_log_file: Option<PathBuf>,
    pub pid: Option<u32>,
}

#[derive(Debug)]
pub struct ServerProcess {
    pub child: Child,
    pub recovered: bool,
}

#[derive(Clone)]
pub struct ConsoleController {
    servers: Arc<Mutex<Vec<Server>>>,
    active_processes: Arc<Mutex<HashMap<String, ServerProcess>>>,
}

impl ConsoleController {
    pub fn new(
        servers: Arc<Mutex<Vec<Server>>>,
        active_processes: Arc<Mutex<HashMap<String, ServerProcess>>>,
    ) -> Self {
        ConsoleController {
            servers,
            active_processes,
        }
    }

    fn get_server(&self, server_id: &str) -> Result<Server, ConsoleError> {
        let servers = self.servers.lock().unwrap();
        servers
            .iter()
            .find(|s| s.id == server_id)
            .cloned()
            .ok_or(ConsoleError::ServerNotFound)
    }

    pub async fn get_live_console(
        &self,
        server_id: &str,
        char_offset: u64,
    ) -> Result<ConsoleResponse, ConsoleError> {
        let server = self.get_server(server_id)?;
        let current_status = server.status.clone();

        // Check server status
        if current_status == "Offline" || current_status == "Stopping" {
            return Ok(ConsoleResponse {
                logs: format!("\n--- Server is {}. Console inactive. ---\n", current_status),
                new_offset: char_offset,
                status: current_status,
            });
        }

        // Check log file configuration
        let log_file = server
            .console_log_file
            .as_ref()
            .ok_or(ConsoleError::LogFileNotConfigured)?;

        if !log_file.exists() {
            if current_status == "Starting" || current_status == "Online" || current_status == "Restarting" {
                return Ok(ConsoleResponse {
                    logs: "--- Console log file not yet created or server not fully started. ---\n".to_string(),
                    new_offset: 0,
                    status: current_status,
                });
            } else {
                return Ok(ConsoleResponse {
                    logs: format!("\n--- Server is {}. Console may be stale. ---\n", current_status),
                    new_offset: char_offset,
                    status: current_status,
                });
            }
        }

        // Get file size
        let metadata = fs::metadata(&log_file).await?;
        let total_size = metadata.len();

        if char_offset >= total_size {
            if current_status != "Starting" && current_status != "Online" && current_status != "Restarting" {
                return Ok(ConsoleResponse {
                    logs: format!("\n--- Server is {}. Console may be stale. ---\n", current_status),
                    new_offset: total_size,
                    status: current_status,
                });
            } else {
                return Ok(ConsoleResponse {
                    logs: String::new(),
                    new_offset: total_size,
                    status: current_status,
                });
            }
        }

        // Read new content
        let mut file = File::open(&log_file).await?;
        file.seek(std::io::SeekFrom::Start(char_offset)).await?;

        let mut buffer = Vec::with_capacity((total_size - char_offset) as usize);
        file.read_to_end(&mut buffer).await?;

        let new_content = String::from_utf8(buffer).map_err(|_| ConsoleError::InternalError)?;

        Ok(ConsoleResponse {
            logs: new_content,
            new_offset: total_size,
            status: current_status,
        })
    }

    pub async fn get_full_log(&self, server_id: &str) -> Result<String, ConsoleError> {
        let server = self.get_server(server_id)?;
        let log_file = server
            .console_log_file
            .as_ref()
            .ok_or(ConsoleError::LogFileNotConfigured)?;

        if !log_file.exists() {
            return Err(ConsoleError::LogFileNotFound);
        }

        let content = fs::read_to_string(&log_file).await?;
        Ok(content)
    }

    pub async fn send_command(
        &self,
        server_id: &str,
        command: &str,
    ) -> Result<(), ConsoleError> {
        // Validate command
        let trimmed = command.trim();
        if trimmed.is_empty() {
            return Err(ConsoleError::InvalidCommand);
        }

        // Check server status
        let server = self.get_server(server_id)?;
        let allowed_statuses = ["Online", "Starting", "Restarting"];
        if !allowed_statuses.contains(&server.status.as_str()) {
            return Err(ConsoleError::InvalidServerState(server.status));
        }

        // Get process
        let mut processes = self.active_processes.lock().unwrap();
        let process = processes
            .get_mut(server_id)
            .ok_or(ConsoleError::ProcessInputUnavailable)?;

        // Check if recovered process
        if process.recovered && process.child.stdin.is_none() {
            return Err(ConsoleError::ProcessInputUnavailable);
        }

        // Send command
        if let Some(stdin) = &mut process.child.stdin {
            stdin
                .write_all(format!("{}\n", trimmed).as_bytes())
                .map_err(|_| ConsoleError::ProcessInputUnavailable)?;
            Ok(())
        } else {
            Err(ConsoleError::ProcessInputUnavailable)
        }
    }
}

// Response types
#[derive(Debug, Serialize)]
pub struct ConsoleResponse {
    pub logs: String,
    pub new_offset: u64,
    pub status: String,
}

// Handler functions for Actix-web
pub async fn handle_live_console(
    console_controller: web::Data<ConsoleController>,
    server_id: web::Path<String>,
    query: web::Query<HashMap<String, String>>,
) -> Result<HttpResponse, ConsoleError> {
    let offset = query
        .get("offset")
        .and_then(|s| s.parse().ok())
        .unwrap_or(0);

    let response = console_controller
        .get_live_console(&server_id, offset)
        .await?;

    Ok(HttpResponse::Ok().json(response))
}

pub async fn handle_full_log(
    console_controller: web::Data<ConsoleController>,
    server_id: web::Path<String>,
) -> Result<HttpResponse, ConsoleError> {
    let content = console_controller.get_full_log(&server_id).await?;
    Ok(HttpResponse::Ok()
        .content_type("text/plain")
        .body(content))
}

pub async fn handle_send_command(
    console_controller: web::Data<ConsoleController>,
    server_id: web::Path<String>,
    payload: web::Json<HashMap<String, String>>,
) -> Result<HttpResponse, ConsoleError> {
    let command = payload
        .get("command")
        .ok_or(ConsoleError::InvalidCommand)?;

    console_controller
        .send_command(&server_id, command)
        .await?;

    Ok(HttpResponse::Ok().json(serde_json::json!({
        "message": format!("Command \"{}\" sent.", command.trim())
    })))
}